python/pyarrow/array.pxi

```diff
@@ -3076,6 +3076,111 @@ cdef class ExtensionArray(Array):
         return Array._to_pandas(self.storage, options, **kwargs)
 
 
+class FixedShapeTensorArray(ExtensionArray):
+    """
+    Concrete class for fixed shape tensor extension arrays.
+
+    Examples
+    --------
+    Define the extension type for tensor array
+
+    >>> import pyarrow as pa
+    >>> tensor_type = pa.fixed_shape_tensor(pa.int32(), [2, 2])
+
+    Create an extension array
+
+    >>> arr = [[1, 2, 3, 4], [10, 20, 30, 40], [100, 200, 300, 400]]
+    >>> storage = pa.array(arr, pa.list_(pa.int32(), 4))
+    >>> pa.ExtensionArray.from_storage(tensor_type, storage)
+    <pyarrow.lib.FixedShapeTensorArray object at ...>
+    [
+      [
+        1,
+        2,
+        3,
+        4
+      ],
+      [
+        10,
+        20,
+        30,
+        40
+      ],
+      [
+        100,
+        200,
+        300,
+        400
+      ]
+    ]
+    """
+
+    def to_numpy_ndarray(self):
+        """
+        Convert fixed shape tensor extension array to a numpy array (with dim+1).
+        """
+        np_flat = np.asarray(self.storage.values)
+        numpy_tensor = np_flat.reshape((len(self),) + tuple(self.type.shape),
+                                       order='C')
```

Comment:
Does reshape incur a copy? Perhaps we should throw if `permutation[0] != 0` (as that would mean array can't be chunked)?
Also shouldn't we take permutation into account when reshaping?

Reply:
It doesn't guarantee to be zero-scopy in all cases, but in this simple case of reshaping a 1D array to an nD array, this is always zero copy AFAIK

Reply:
What happens in cases where permutation doesn't give C style layout?

Reply:
A permutation would happen after the reshape, I think? (so the initial reshape is always C contiguous) And that is left to the user for now in this method (which we should document)

Reply:
Right, but permutation would not roundtrip to numpy object and back. I'd prefer to throw NotImplemented if permutation is not trivial (C).

Reply:
Raising an error for now sounds fine to me

Reply:
Great! I think raising an error (or a warning?) sounds good to me too 👍 
Will add it in `to_numpy_ndarray` for cases when `permutation` is not trivial (`[0, 1, ... N-1]`). Do we want to add a warning for `dim_names not None` also?

Reply:
Ah, actually I need to error if `permutation[0] != 0`. And if the permutation is for example `[0, 2, 1]` the user could still use the information about the permutation to rearrange the ordering after the reshape in `to_numpy_ndarray`. Please correct me if I am wrong =)

Reply:
> Ah, actually I need to error if `permutation[0] != 0`.

No, the `self.type.permutation` is only that of the individual tensor elements (given that you have a FixedShapeTensorArray, by definition the first dimension of the n+1 dim ndarray is always the length of the array, and can't be permuted)

> And if the permutation is for example `[0, 2, 1]` the user could still use the information about the permutation to rearrange the ordering after the reshape in `to_numpy_ndarray`.

Yes, and as mentioned before, I think we should certainly add examples how the user can do this (if we decide to not automatically do it in `to_numpy_ndarray`)

> Do we want to add a warning for dim_names not None also?

I wouldn't raise a warning for that, since it's a fact of life that numpy arrays don't have dimension names, so that's just a consequence of calling this method. We can mention that in the docstring, though, to be explicit.

Reply:
> No, the `self.type.permutation` is only that of the individual tensor elements (given that you have a FixedShapeTensorArray, by definition the first dimension of the n+1 dim ndarray is always the length of the array, and can't be permuted)

Yes, thank you for confirming the thought I was currently struggling with!
Then I do not see the need for raising an error (as it would be raised for every permutation that is not `None`) but maybe make it explicit in the docstrings that the "0" dimension is by default not permutable and is fixed (see the examples and description I added in the docs PR for this binding: https://github.com/apache/arrow/pull/34957/files)

> Yes, and as mentioned before, I think we should certainly add examples how the user can do this (if we decide to not automatically do it in `to_numpy_ndarray`)

Done: https://github.com/apache/arrow/pull/34957/files

> I wouldn't raise a warning for that, since it's a fact of life that numpy arrays don't have dimension names, so that's just a consequence of calling this method. We can mention that in the docstring, though, to be explicit.

Agree, will make it explicit in the docstrings.

Reply:
After couple of days to think it over, I would agree with all the comments from Joris at the beginning of the thread.

1. The use of permutation to reorder data should (for now) be left to the user to decide if they want to do that or not. I have added a note in the docstrings: https://github.com/apache/arrow/pull/34883/commits/dd8fd31f4b15ca2e0a39673ab2a338b991cd2032
2. `to_numpy_ndarray` is reshaping 1-D array so there is no need to check the layout style as 1-D array is always both C-contiguous and F-contiguous.
3. We cannot create a F-contiguous tensor in PyArrow and there is a check for row layout in `from_numpy_ndarray` so there is no option, currently, to have a F-contiguous tensors in the tensor array anyways.

With this I propose to leave the code as is and maybe open an issue later to discuss changes/features further.

I have also added a PR for the documentation of the `fixed_shape_tensor` binding here: https://github.com/apache/arrow/pull/34957. After the binding PR is merged I will run the code again, build the docs locally to check the html and then mark the PR as ready for review.

Reply:
> 2. `to_numpy_ndarray` is reshaping 1-D array so there is no need to check the layout style as 1-D array is always both C-contiguous and F-contiguous.

While that's true for the physical layout it's not necessarily true for the logical layout. I propose (https://github.com/apache/arrow/pull/34883#pullrequestreview-1378208716) we block converting tensors with non-trivial permutations and add correct logic to handle those later. (Another option is to go via `array.ToTensor().to_numpy()` once [`FromTensor/ToTensor`](https://github.com/apache/arrow/pull/34797) is merged.)